tools/c7n_azure/tests/test_tags.py

```diff
@@ -424,4 +427,204 @@ def test_auto_tag_days_must_be_btwn_1_and_90(self):
+                 }
+            ],
+        })
+        p.run()
+
+        # verify that tags were trimmed to
+        # have 14 spaces and 1 preserved
+        vm = client.virtual_machines.get('test_vm', 'cctestvm')
+        self.assertEqual(len(vm.tags), 1)
+
+    @arm_template('vm.json')
+    def test_tag_trim_space_0_removes_all_tags_but_preserve(self):
+        """Verifies tag trim removes all other tags but tags
+        listed in preserve
+        """
+
+        # Add tags to trim
+        p = self.load_policy({
+            'name': 'test-azure-tag',
+            'resource': 'azure.vm',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'eq',
+                 'value_type': 'normalize',
+                 'value': 'cctestvm'}
+            ],
+            'actions': [
+                {'type': 'tag',
+                 'tags': {'tag-to-trim1': 'value1', 'tag-to-trim2': 'value2',
+                          'tag-to-trim3': 'value3'}}
+            ],
+        })
+        p.run()
+
+        # verify initial tags contain more than testtag
+        s = Session()
+        client = s.client('azure.mgmt.compute.ComputeManagementClient')
+        vm = client.virtual_machines.get('test_vm', 'cctestvm')
+        self.assertTrue('tag-to-trim1' in vm.tags)
+        self.assertTrue('tag-to-trim2' in vm.tags)
+        self.assertTrue('tag-to-trim3' in vm.tags)
+        self.assertTrue('testtag' in vm.tags)
+
+        p = self.load_policy({
+            'name': 'test-azure-tag',
+            'resource': 'azure.vm',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'eq',
+                 'value_type': 'normalize',
+                 'value': 'cctestvm'}
+            ],
+            'actions': [
+                {'type': 'tag-trim',
+                 'space': 0,
+                 'preserve': ['testtag']
+                 }
+            ],
+        })
+        p.run()
+
+        # verify all tags trimmed but testtag
+        vm = client.virtual_machines.get('test_vm', 'cctestvm')
+        self.assertEqual(vm.tags, {'testtag': 'testvalue'})
+
+    @arm_template('vm.json')
+    @patch('logging.Logger.warning')
+    def test_tag_trim_warns_no_candidates(self, logger_mock):
+        """Verifies tag trim warns when there are no candidates
+        to trim
+        """
+
+        p = self.load_policy({
+            'name': 'test-azure-tag',
+            'resource': 'azure.vm',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'eq',
+                 'value_type': 'normalize',
+                 'value': 'cctestvm'}
+            ],
+            'actions': [
+                {'type': 'tag-trim',
+                 'space': 0,
+                 'preserve': ['testtag']
+                 }
+            ],
+        })
+        p.run()
+
+        logger_mock.assert_called()
```

Comment:
Is it possible to get a more detailed assert; I don't want this test to have false positives? I suppose the patch for 'warning' might be sufficient.

Reply:
Yeah I could do called_with but it's going to be called with a long interpolated string with the resource's id, but I'll see if I can make it more robust.